### PR TITLE
Fix dtypes related to floating point precision

### PR DESCRIPTION
**Pull request by Scienfitz**
Since floating point precision can be controlled via env vars (#226) various problems have surfaced letting tests fail in single precision.
This PR fixes those. They were mostly related to the way `values` and `comp_df` were created for parameters, `selection` was treated in `SubSelectionCondition` and a `lookup` in a different float precision being used in a simulation.

The only remaining issues with test in single precision are numerical instabilities (out of scope)

### CODE REVIEW

File: baybe/constraints/base.py [hunk 1]
```diff
@@ -173,11 +175,10 @@ def to_botorch(
             if p in param_names
         ]
 
-        # TODO: Cast rhs to correct precision once BoTorch also supports single point.
         return (
             torch.tensor(param_indices),
             torch.tensor(self.coefficients, dtype=DTypeFloatTorch),
-            self.rhs,
+            np.asarray(self.rhs, dtype=DTypeFloatNumpy).item(),
```
Comment (AdrianSosic) [inline baybe/constraints/base.py:181]: I think this is rather cumbersome. Why not simply `DTypeFloatNumpy(self.rhs)`?

File: baybe/utils/chemistry.py [hunk 2]
```diff
@@ -88,13 +88,10 @@ def _smiles_to_mordred_features(smiles: str) -> np.ndarray:
     """
     try:
         return np.asarray(
-            _mordred_calculator(Chem.MolFromSmiles(smiles)).fill_missing(),
-            dtype=DTypeFloatNumpy,
+            _mordred_calculator(Chem.MolFromSmiles(smiles)).fill_missing()
```
Comment (AdrianSosic) [inline baybe/utils/chemistry.py:91]: Just for me to understand: your rational is to say "conversion only happens in the baybe objects, not utilities"?

File: baybe/parameters/categorical.py [hunk 3]
```diff
@@ -47,9 +48,13 @@ def comp_df(self) -> pd.DataFrame:  # noqa: D102
         # See base class.
         if self.encoding is CategoricalEncoding.OHE:
             cols = [f"{self.name}_{val}" for val in self.values]
-            comp_df = pd.DataFrame(np.eye(len(self.values), dtype=int), columns=cols)
+            comp_df = pd.DataFrame(
```
Comment (AdrianSosic) [inline baybe/parameters/categorical.py:51]: So I guess this is the part you mentioned in the chat, right? Seems reasonable, but I'm also not sure where is the best place for the cast. In principle, one could also use bool/int type here and then only cast when constructing the searchspace comp_df (because there, there whole array must share a single dtype), but I'm not sure if anything is gained by that...

File: baybe/parameters/numerical.py [hunk 4]
```diff
@@ -79,12 +79,14 @@ def _validate_tolerance(  # noqa: DOC101, DOC103
     @property
     def values(self) -> tuple:  # noqa: D102
         # See base class.
-        return self._values
+        return tuple(DTypeFloatNumpy(itm) for itm in self._values)
```
Comment (AdrianSosic) [inline baybe/parameters/numerical.py:82]: Asking without yet having a strong preference: is there a reason why you decided to cast in the property, instead of casting in the converter (i.e. when assigning)? I also guess that your way is probably the better alternative, but wanted to discuss nevertheless

**Review by AdrianSosic — COMMENTED** [review]
Hi @Scienfitz, thanks for the PR. Here a few comments to discuss

File: baybe/parameters/categorical.py (hunk 3 above)
Comment (Scienfitz) [reply baybe/parameters/categorical.py:51]: hmm id rather have specified types as close to the "source" as possible, imo it would be strange if `comp_df` had other types than the resulting `comp_rep`
Not sure if the `comp_df` is used anywhere else but issues could be coming up i we shift the type responsibility to the product utility

File: baybe/utils/chemistry.py (hunk 2 above)
Comment (Scienfitz) [reply baybe/utils/chemistry.py:91]: no in this file I shifted the DTYPE conversion to the next function below which sues this one here, strangely when the conversion was done here it did not always have correct type
this delete is just there because the other function without the underscore does the conversion now

File: baybe/constraints/base.py (hunk 1 above)
Comment (Scienfitz) [reply baybe/constraints/base.py:181]: mypy (complaining you suggestion returns `object`)
happy to sue an alternative if you want (cast? ignore?) let me know

File: CHANGELOG.md [hunk 5]
```diff
@@ -11,6 +11,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 - Non-GP surrogates not working with `deepcopy` and the simulation module due to slotted
   base class
+- Datatype inconsistencies for various parameter's `values` and `comp_df` related to
```
[the remark this answers is not in the document]
Comment (Scienfitz) [reply CHANGELOG.md:14]: yep fixed

File: baybe/utils/chemistry.py (hunk 2 above)
Comment (AdrianSosic) [reply baybe/utils/chemistry.py:91]: But that's what I meant: if someone were to use the utility directly, we say we don't bother about dtypes, only when used in "actual" baybe context. Have no strong opinion on what is better though...

File: baybe/constraints/base.py (hunk 1 above)
Comment (AdrianSosic) [reply baybe/constraints/base.py:181]: oh, OK, that's unexpected and bad. Then I have no strong preference here, proceed as you wish :D

File: baybe/parameters/categorical.py (hunk 3 above)
Comment (AdrianSosic) [reply baybe/parameters/categorical.py:51]: Yeah, probably too much of a hassle. Happy to keep it as is 👍🏼

File: baybe/utils/chemistry.py (hunk 2 above)
Comment (Scienfitz) [reply baybe/utils/chemistry.py:91]: hmm didnt think about that, the utility that is now potentially not in the right dtype is the private one with `_` anyway, the none private one has the right dtype so imo this is reasonable

File: baybe/parameters/numerical.py (hunk 4 above)
Comment (Scienfitz) [reply baybe/parameters/numerical.py:82]: good point didnt think about that i think it would be somewhat nicer int he converter
is ti possible to do it like this `converter=lambda x: sorted(cattrs.structure(x, tuple[DTypeFloatNumpy, ...]))` ?

Comment (AdrianSosic) [reply baybe/parameters/numerical.py:82]: In fact, there was a similar reason why I did not normalize the weights in the desirability objective during initialization, but rather kept the raw input. You remember?

Comment (Scienfitz) [reply baybe/parameters/numerical.py:82]: yes rememebr and it make sense, despite being prob a little more elegant thats an arg against the converter, so I can keep it and resolve?

Comment (AdrianSosic) [reply baybe/parameters/numerical.py:82]: Yeah, let's keep it. Closing ...

File: baybe/utils/chemistry.py (hunk 2 above)
Comment (AdrianSosic) [reply baybe/utils/chemistry.py:91]: OK, I guess there is no obvious answer anyway, so let's just keep until we run into problems 👍🏼

File: tests/hypothesis_strategies/basic.py [hunk 6]
```diff
@@ -3,6 +3,14 @@
...
 
-finite_floats = partial(st.floats, allow_infinity=False, allow_nan=False)
+from baybe.utils.numerical import DTypeFloatNumpy
+
+finite_floats = partial(
+    st.floats,
+    allow_infinity=False,
+    allow_nan=False,
+    width=32 if DTypeFloatNumpy == np.float32 else 64,
+)
```
Comment (AdrianSosic) [inline tests/hypothesis_strategies/basic.py:15]: I think this part here perhaps mixes two properties that should rather stay independent – because we want the float32/64 control even for non-finite floats. Therefore, perhaps it makes more sense to change this into
```suggestion
dtype_floats = partial(st.floats, width=32 if DTypeFloatNumpy == np.float32 else 64)
"""A strategy producing floats of the requested dtype."""

finite_floats = partial(dtype_floats, allow_infinity=False, allow_nan=False)
```
And then really use `dtype_floats` in placed where we only want to control the type?

**Review by AdrianSosic — COMMENTED** [review]
Last few comments, I hope 🙃

File: baybe/constraints/conditions.py [hunk 7]
```diff
@@ -135,9 +146,39 @@ class SubSelectionCondition(Condition):
     """Class for defining valid parameter entries."""
 
     # object variables
-    selection: list[Any] = field()
-    """The list of items which are considered valid."""
+    _selection: tuple = field(
+        # FIXME[typing]: https://github.com/python-attrs/cattrs/issues/111
+        converter=lambda x: cattrs.structure(x, tuple),  # type: ignore
```
[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/constraints/conditions.py:151]: changed

File: baybe/simulation/lookup.py [hunk 8]
```diff
@@ -79,9 +80,13 @@ def _look_up_target_values(
             queries[target.name] = measured_targets.iloc[:, k_target]
 
     # Get results via dataframe lookup (works only for exact matches)
-    # IMPROVE: Although its not too important for a simulation, this
+    # IMPROVE: Although it's not too important for a simulation, this
     #  could also be implemented for approximate matches
     elif isinstance(lookup, pd.DataFrame):
+        # Enforce correct float precision in the lookup dataframe
+        float_cols = lookup.select_dtypes(include=["float"]).columns
+        lookup[float_cols] = lookup[float_cols].astype(DTypeFloatNumpy)
```
[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/simulation/lookup.py:88]: ok changed it and also moved to `simulate_experiment` where it creates a copy before modifying, this should also do the modifications fewer times which possibly affects speed

**Review by AdrianSosic — APPROVED** [review]
